include/crypto/ml_kem.h

```diff
@@ -0,0 +1,475 @@
+ * generation consumes two random values (d, z) with "d" plus the rank (domain
+ * separation) further expanded to two derived seeds "rho" and "sigma", with
+ * "rho" used to generate the public matrix "A", and sigma to generate the
+ * private vector "s" and error vector "e".
+ *
+ * Encapsulation also consumes one random value m, that is 32-bytes long.  The
+ * resulting shared secret "K" (also 32 bytes) and an internal random value "r"
+ * are derived from "m" concatenated with a digest of the received public key.
+ * Use of the public key hash means that the derived shared secret is
+ * "contributary", it uses randomness from both parties.
+ *
+ * The seed "rho" is appended to the public key and allows the recipient of the
+ * public key to re-compute the matrix "A" when performing encapsulation.
+ *
+ * Note that the matrix "m" we store in the public key is the transpose of the
+ * "A" matrix from FIPS 203!
+ */
+#  define ML_KEM_RANDOM_BYTES    32 /* rho, sigma, ... */
+#  define ML_KEM_SEED_BYTES      (ML_KEM_RANDOM_BYTES * 2) /* Keygen (d, z) */
+
+#  define ML_KEM_PKHASH_BYTES        32 /* Salts the shared-secret */
+#  define ML_KEM_SHARED_SECRET_BYTES 32
+
+/*-
+ * The ML-KEM specification can be found in
+ * https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.pdf
+ *
+ * Section 8, Table 2, lists the parameters for the three variants:
+ *
+ *         Variant     n     q  k  eta1  eta2  du   dv  secbits
+ *      ----------   ---  ----  -  ----  ----  --   --  -------
+ *      ML-KEM-512   256  3329  2     3     2  10    4      128
+ *      ML-KEM-768   256  3329  3     2     2  10    4      192
+ *      ML-KEM-1024  256  3329  4     2     2  11    5      256
+ *
+ * where:
+ *
+ * - "n" (ML_KEM_DEGREE above) is the fixed degree of the quotient polynomial
+ *    in the ring: "R_q" = Z[X]/(X^n + 1).
+ * - "q" (ML_KEM_PRIME above) is the fixed prime (256 * 13 + 1 = 3329) used in
+ *   all ML-KEM variants.
+ * - "k" is the row rank of the square matrix "A", with entries in R_q, that
+ *   defines the "noisy" linear equations: t = A * s + e.  Also the rank of
+ *   of the associated vectors.
+ * - "eta1" determines the amplitude of "s" and "e" vectors in key generation
+ *   and the "y" vector in ML-KEM encapsulation (K-PKE encryption).
+ * - "eta2" determines the amplitude of "e1" and "e2" noise terms in ML-KEM
+ *   encapsulation (K-PKE encryption).
+ * - "du" determines how many bits of each coefficient are retained in the
+ *   compressed form of the "u" vector in the encapsulation ciphertext.
+ * - "dv" determines how many bits of each coefficient are retained in the
+ *   compressed form of the "v" value in encapsulation ciphertext
+ * - "secbits" is required security strength of the RNG for the random inputs.
+ */
+
+/*
+ * Variant-specific CBD vector generation helpers, these generate a single CBD
+ * scalar.
+ */
+typedef __owur
+int ossl_ml_kem_cbd_func(ossl_ml_kem_scalar *out,
```

Comment:
should be uppercase

Reply:
Now private to the single C-source file in which It is used.